src/chan_vese.jl

```diff
@@ -0,0 +1,241 @@
+    N = ndims(img)
+    iter = 0
+    h = 1.0
+    del = tol + 1
+    img .= img .- minimum(img)
+
+    if maximum(img) != 0
+        img .= img ./ maximum(img)
+    end
+
+    # Precalculation of some constants which helps simplify some integration   
+    # area = length(img) # area = ∫H𝚽 + ∫H𝚽ⁱ
+    area = length(img) # area = ∫H𝚽 + ∫H𝚽ⁱ
+    ∫u₀ = sum(img)     # ∫u₀ = ∫u₀H𝚽 + ∫u₀H𝚽ⁱ
+
+    # Initialize the level set
+    𝚽ⁿ = initial_level_set(size(img))
+
+    # Preallocation and initializtion
+    H𝚽 = trues(size(img)...)
+    𝚽ⁿ⁺¹ = similar(𝚽ⁿ)
+
+    Δ = ntuple(d -> CartesianIndex(ntuple(i -> i == d ? 1 : 0, N)), N)
+    idx_first = first(CartesianIndices(𝚽ⁿ))
+    idx_last = last(CartesianIndices(𝚽ⁿ))
+    
+    while (del > tol) & (iter < max_iter)
+        ϵ = 1e-8
+        diff = 0
+
+        # Calculate the average intensities
+        @. H𝚽 = 𝚽ⁿ > 0 # Heaviside function
+        c₁, c₂ = calculate_averages(img, H𝚽, area, ∫u₀) # Compute c₁(𝚽ⁿ), c₂(𝚽ⁿ)
+
+        # Calculate the variation of level set 𝚽ⁿ
+        @inbounds @simd for idx in CartesianIndices(𝚽ⁿ)
+            𝚽₀  = 𝚽ⁿ[idx] # 𝚽ⁿ(x, y)
+            u₀ = img[idx]  # u₀(x, y)
+            𝚽₊ = broadcast(i->𝚽ⁿ[i], ntuple(d -> idx[d] != idx_last[d]  ? idx + Δ[d] : idx, N))
+            𝚽₋ = broadcast(i->𝚽ⁿ[i], ntuple(d -> idx[d] != idx_first[d] ? idx - Δ[d] : idx, N))
+
+            # Solve the PDE of equation 9 in paper[3]
+            center_diff = ntuple(d -> (𝚽₊[d] - 𝚽₋[d])^2 / 4., N)
+            sum_center_diff = sum(center_diff)
+            C₊ = ntuple(d -> 1. / sqrt(ϵ + (𝚽₊[d] - 𝚽₀)^2 + sum_center_diff - center_diff[d]), N)
+            C₋ = ntuple(d -> 1. / sqrt(ϵ + (𝚽₋[d] - 𝚽₀)^2 + sum_center_diff - center_diff[d]), N)
+
+            K = sum(𝚽₊ .* C₊) + sum(𝚽₋ .* C₋)
+            δₕ = h / (h^2 + 𝚽₀^2) # Regularised Dirac function
+            difference_from_average = - λ₁ * (u₀ - c₁) ^ 2 + λ₂ * (u₀ - c₂) ^ 2
+
+            𝚽ⁿ⁺¹[idx] = 𝚽 = (𝚽₀ + Δt * δₕ * (μ * K + difference_from_average)) / (1. + μ * Δt * δₕ * (sum(C₊) + sum(C₋)))
+            diff += (𝚽 - 𝚽₀)^2
+        end
+
+        del = sqrt(diff / area)
+
+        if reinitial_flag
+            # Reinitialize 𝚽 to be the signed distance function to its zero level set
+            reinitialize!(𝚽ⁿ⁺¹, 𝚽ⁿ, Δt, h)
+        else
+            𝚽ⁿ .= 𝚽ⁿ⁺¹
+        end
+  
+        iter += 1
+    end
+
+    return 𝚽ⁿ .> 0
+end
+
+function initial_level_set(shape::Tuple{Int64, Int64})
+    x₀ = reshape(collect(0:shape[begin]-1), shape[begin], 1)
+    y₀ = reshape(collect(0:shape[begin+1]-1), 1, shape[begin+1])
+    𝚽₀ = @. sin(pi / 5 * x₀) * sin(pi / 5 * y₀)
+end
+
+function initial_level_set(shape::Tuple{Int64, Int64, Int64})
+    x₀ = reshape(collect(0:shape[begin]-1), shape[begin], 1, 1)
+    y₀ = reshape(collect(0:shape[begin+1]-1), 1, shape[begin+1], 1)
+    z₀ = reshape(collect(0:shape[begin+2]-1), 1, 1, shape[begin+2])
+    𝚽₀ = @. sin(pi / 5 * x₀) * sin(pi / 5 * y₀) * sin(pi / 5 * z₀)
+end
```

Comment:
```suggestion
initial_level_set(sz) = broadcast(*, kernelfactors(ntuple(i -> sin.((pi / 5) * (0:sz[i]-1)), length(sz)))
```
`kernelfactors` is from ImageFiltering and will make this truly arbitrary-dimensional.

Reply:
I’ve tried to change the code, but do I need to change the code like following to get a `Matrix` type output ?
```julia
initial_level_set(sz) = broadcast(*, kernelfactors(ntuple(i -> sin.((pi / 5) * (0:sz[i]-1)), length(sz)))...)
```
Another question is that this changing seems reduce the performance:
```julia
# previous code's performance
function initial_level_set(shape::Tuple{Int64, Int64})
    x₀ = reshape(collect(0:shape[begin]-1), shape[begin], 1)
    y₀ = reshape(collect(0:shape[begin+1]-1), 1, shape[begin+1])
    𝚽₀ = @. sin(pi / 5 * x₀) * sin(pi / 5 * y₀)
end
...
julia> @btime chan_vese(img_gray, μ=0.25, λ₁=1.0, λ₂=1.0, tol=1e-3, max_iter=200, Δt=0.5, reinitial_flag=false);
  462.876 ms (119 allocations: 6.46 MiB)

# using kernelfactor
initial_level_set(sz) = broadcast(*, kernelfactors(ntuple(i -> sin.((pi / 5) * (0:sz[i]-1)), length(sz)))...)
...
julia> @btime chan_vese(img_gray, μ=0.25, λ₁=1.0, λ₂=1.0, tol=1e-3, max_iter=200, Δt=0.5, reinitial_flag=false);
  575.990 ms (115 allocations: 6.46 MiB)
```
It seems that the changing causes more than `100ms`'s performance reduction.

Reply:
Shoot, that should have been
```julia
initial_level_set(sz) = broadcast(*, kernelfactors(ntuple(i -> sin.((pi / 5) * (0:sz[i]-1)), length(sz)))...)
```
so that
```julia
julia> initial_level_set((3, 5))
3×5 OffsetArray(::Matrix{Float64}, 1:3, 1:5) with eltype Float64 with indices 1:3×1:5:
 0.0  0.0       0.0       0.0       0.0
 0.0  0.345492  0.559017  0.559017  0.345492
 0.0  0.559017  0.904508  0.904508  0.559017

julia> initial_level_set((3, 5, 2))
3×5×2 OffsetArray(::Array{Float64, 3}, 1:3, 1:5, 1:2) with eltype Float64 with indices 1:3×1:5×1:2:
[:, :, 1] =
 0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  0.0

[:, :, 2] =
 0.0  0.0       0.0       0.0       0.0
 0.0  0.203075  0.328582  0.328582  0.203075
 0.0  0.328582  0.531657  0.531657  0.328582
```
Sorry about that.